### PR TITLE
Make torch examples to use ray

### PR DESCRIPTION
**Pull request by mortid0**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

Previous Behavior
----------------
Torch examples were using deepgnn-torch package.

New Behavior
----------------
* Examples are self contained and tested E2E via bazel.
* Removed link prediction model, because they were not tested anyway.
* Added TGN model.
* Added GCN model with distributed training.

### CODE REVIEW

File: examples/pytorch/gat.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        scores = self.conv2(x, edge_index)
+        labels = labels[mask]  # [batch_size]
+        scores = scores[mask]  # [batch_size]
+        pred = scores.argmax(dim=1)
+        loss = self.xent(scores, labels)
+
+        return loss, pred, labels
+
+
+def _train_func(config: dict):
```
Comment (coledie) [inline examples/pytorch/gat.py:158]: Why _ it is not a hidden function?

File: examples/pytorch/gat.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        if worker_info is None:
+            return map(self.query, batches)
+
+        # split workload for distributed training
+        return map(self.query, batches[worker_info.id :: worker_info.num_workers])
+
+    def query(self, inputs: np.ndarray) -> tuple:
+        """Fetch training data."""
+        nodes, edges, src_idx = graph_ops.sub_graph(
```
Comment (coledie) [inline examples/pytorch/gat.py:82]: You should test this example on different datasets, last I remember this sub_graph method breaks on all but one dataset.

File: examples/pytorch/gcn.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        num_classes: int,
+        shared_weights: bool = True,
+        dropout: float = 0.0,
+    ):
+        """Initialize model."""
+        super().__init__()
+
+        self.lins = nn.ModuleList()
+        self.lins.append(nn.Linear(feature_dim, hidden_channels))
+        self.lins.append(nn.Linear(hidden_channels, num_classes))
```
Comment (coledie) [inline examples/pytorch/gcn.py:98]: Can pull these appends into the constructor of the modulelist.

File: examples/pytorch/gcn.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        self.dropout = dropout
+        self.convs = nn.ModuleList()
+        for layer in range(num_layers):
```
Comment (coledie) [inline examples/pytorch/gcn.py:101]: same here

File: examples/pytorch/hetgnn/BUILD [hunk 5]
```diff
@@ -1,78 +1,34 @@
...
-        requirement("scikit-learn"),
         requirement("torch"),
-        requirement("networkx"),
-        requirement("tensorboard"),
-        requirement("transformers"),
-        requirement("opencensus"),
-        requirement("opencensus-context"),
-        requirement("opencensus-ext-azure"),
-        requirement("azure-datalake-store"),
+        # requirement("torch-cluster"),
```
Comment (coledie) [inline examples/pytorch/hetgnn/BUILD:28]: Remove commented

File: examples/pytorch/hetgnn/evaluation.py [hunk 6]
```diff
@@ -12,16 +12,16 @@ def load_data(data_file_name, n_features, n_samples):
         data_file = csv.reader(f)
         data = numpy.empty((n_samples, n_features))
```
Comment (coledie) [inline examples/pytorch/hetgnn/evaluation.py:13]: Import numpy as np then replace numpy. with np., this way this code can be better copied into other people's code.

File: examples/pytorch/hetgnn/evaluation.py [hunk 7]
```diff
@@ -30,15 +30,15 @@ def evaluate_node_classification(args, train_num, test_num, datafile_dir):
     train_features = None
     train_target = None
 
-    test_data_f = str(datafile_dir.join("test_class_feature.txt"))
-    test_data = load_data(test_data_f, args.embed_d + 2, test_num)
+    test_data_f = str(config["data_dir"] + ("/test_class_feature.txt"))
```
Comment (coledie) [inline examples/pytorch/hetgnn/evaluation.py:33]: The / in test_class_feature is not cross platform, should be os.path.join(a, b). Applies other spots as well.

File: examples/pytorch/hetgnn/graph.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+import numpy as np
+from torch.utils.data import IterableDataset
+
+from deepgnn.graph_engine import Graph, SamplingStrategy
+
+_node_base_index = 1000000
+
+
+class MockGraph(Graph):
```
Comment (coledie) [inline examples/pytorch/hetgnn/graph.py:15]: Rename file to mockgraph.py?

File: examples/pytorch/hetgnn/model.py [hunk 9]
```diff
@@ -24,20 +22,19 @@ def __init__(
...
+        self.feature_enc = None
+        # )
 
         self.embed_d = embed_d
         self.node_type_count = node_type_count
         self.neighbor_count = neighbor_count
-        self.metric = metric
 
         # NOTE: do not use python list to store nn layers, use nn.ModuleList, otherwise
         # the parameters will not added to the optimizer by default and loss function won't work.
```
Comment (coledie) [inline examples/pytorch/hetgnn/model.py:40]: Maybe remove comments if not needed.

File: examples/pytorch/tgn.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    memory_dim,
+    time_dim,
+    message_module=IdentityMessage(graph.edge_feature_dim, memory_dim, time_dim),
+    aggregator_module=LastAggregator(),
+).to(device)
+
+gnn = GraphAttentionEmbedding(
+    in_channels=memory_dim,
+    out_channels=embedding_dim,
+    msg_dim=graph.edge_feature_dim,
```
Comment (coledie) [inline examples/pytorch/tgn.py:108]: This code belongs in an if __name__ == '__main__', not in file

File: examples/pytorch/tgn.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+logger = get_logger()
+# We can train the model for 2 epochs to get a decent performance.
+for epoch in range(1, 3):
+    loss = train()
+    logger.info(f"Epoch #{epoch:02d}: loss:{loss:.4f}")
+
+test_ap, test_auc = test(test_loader)
+logger.info(f"Average precision{test_ap:.4f}: AUC:{test_auc:.4f}")
+assert test_ap >= 0.72
+assert test_auc >= 0.76
```
Comment (coledie) [inline examples/pytorch/tgn.py:252]: Same here

**Review by coledie — CHANGES_REQUESTED** [review]

File: examples/pytorch/gat.py (hunk 2 above)
Comment (mortid0) [reply examples/pytorch/gat.py:82]: This is an example for specific dataset, typically [other datasets have their own examples](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup_ppi.py).

File: examples/pytorch/gat.py (hunk 1 above)
Comment (mortid0) [reply examples/pytorch/gat.py:158]: to avoid adding an obvious docstring

File: examples/pytorch/gcn.py (hunk 3 above)
Comment (mortid0) [reply examples/pytorch/gcn.py]: Sure, it makes everything longer.

File: examples/pytorch/hetgnn/evaluation.py (hunk 6 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/evaluation.py]: That's the original code..

File: examples/pytorch/hetgnn/graph.py (hunk 8 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/graph.py:15]: That's not java. mockgraph.MockGraph stutters.

File: examples/pytorch/hetgnn/model.py (hunk 9 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/model.py]: yeah, they are really old.

File: examples/pytorch/tgn.py (hunk 10 above)
Comment (mortid0) [reply examples/pytorch/tgn.py]: There are way too many moving parts here and putting everything under __main__ makes code way longer.

File: examples/pytorch/gcn.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        if epoch % 10 == 0:
+            model.eval()
+            ys = labels[masks["test"]]
+            preds = out[masks["test"]].argmax(dim=-1).cpu()
+            accuracy = int((ys == preds).sum()) / int(len(ys))
+            get_logger().info(
+                f"Epoch: {epoch:>2d}; Loss: {(total_loss/num_examples):>4.2f}; Accuracy: {accuracy:.4f};"
+            )
+
+    assert accuracy > 0.45
```
Comment (coledie) [inline examples/pytorch/gcn.py:193]: Is 45% accuracy the best we can do?

File: examples/pytorch/hetgnn/main.py [hunk 13]
```diff
@@ -1,90 +1,672 @@
...
+                                    int(p_a_list[t][i][k][1:])
+                                )
+                            if (
+                                int(p_a_list[t][i][j][1:])
+                                not in a_a_list_train[int(p_a_list[t][i][k][1:])]
+                            ):
+                                a_a_list_test[int(p_a_list[t][i][k][1:])].append(
+                                    int(p_a_list[t][i][j][1:])
+                                )
+
```
Comment (coledie) [inline examples/pytorch/hetgnn/main.py:252]: We need more documentation for what each variable contains since they are all 4 dimensional, also docs on what each file is for // rename file to make it clear. Also docstrings would help. Otherwise it is not very useful to have an example that no one can read.

Comment (coledie) [reply examples/pytorch/hetgnn/main.py:252]: Also what is the need for this example that is not already covered by other examples? It seems not very useful to have such a complex example.

**Review by coledie — CHANGES_REQUESTED** [review]

File: examples/pytorch/tgn.py (hunk 10 above)
Comment (coledie) [reply examples/pytorch/tgn.py]: How so? You can put train/test functions in main if they use the globals

Comment (mortid0) [reply examples/pytorch/tgn.py]: you don't need check module name then if you want to keep globals.

File: examples/pytorch/hetgnn/main.py (hunk 13 above)
Comment (mortid0) [reply examples/pytorch/hetgnn/main.py:252]: This is outside of scope of this PR. I made this example self contained without changing training code.

File: examples/pytorch/gcn.py (hunk 12 above)
Comment (mortid0) [reply examples/pytorch/gcn.py:193]: No, this is a good enough number for training for 100 epochs under 30 seconds.

**Review by coledie — APPROVED** [review]
